[clang][analyzer] Add StoreToImmutable checker

gamesh411 opened the pull request:
This adds alpha.core.StoreToImmutable, a new alpha checker that detects writes
to immutable memory regions, implementing part of SEI CERT Rule ENV30-C. The
original proposal only handled global const variables, but this implementation
extends it to also detect writes to:
- Local const variables
- String literals
- Const parameters and struct members
- Const arrays and pointers to const data

This checker is the continuation of the work started by zukatsinadze.
Discussion: https://reviews.llvm.org/D124244

gamesh411 committed:
[clang][analyzer] Add StoreToImmutable checker

This adds alpha.core.StoreToImmutable, a new alpha checker that detects
writes
to immutable memory regions, implementing part of SEI CERT Rule ENV30-C.
The
original proposal only handled global const variables, but this
implementation
extends it to detect writes to:
- Global const variables
- Local const variables
- String literals
- Const parameters and struct members
- Const arrays and pointers to const data

This checker is the continuation of the work started by @zukatsinadze.
Discussion: https://reviews.llvm.org/D124244

steakhal commented on clang/lib/StaticAnalyzer/Checkers/StoreToImmutableChecker.cpp:
```diff
@@ -0,0 +1,146 @@
+//=== StoreToImmutableChecker.cpp - Store to immutable memory checker -*- C++
+//-*-===//
```
This is broken into two lines.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/StoreToImmutableChecker.cpp:
```diff
@@ -0,0 +1,146 @@
+};
+} // end anonymous namespace
+
+bool StoreToImmutableChecker::isConstVariable(const MemRegion *MR,
+                                              CheckerContext &C) const {
```
Could you please paste the test coverage of this function? It has a lot of branches, and I'm not sure if all of these are feasible.

steakhal commented on clang/test/Analysis/store-to-immutable-basic.cpp:
```diff
@@ -0,0 +1,166 @@
+void test_direct_write_to_const_global() {
+  // This should trigger a warning about writing to immutable memory
+  *(int*)&global_const = 100; // expected-warning {{Writing to immutable memory is undefined behavior}}
+  // expected-note@-1 {{Writing to immutable memory is undefined behavior. This memory region is marked as immutable and should not be modified}}
```
By the looks of it these `expected-notes` are dead, as the text diagnostic output is not enabled in the RUN line.
I bet if you would remove these, things would still pass.
Since the checked does not have any state or uses any values we don't need to check the bug report notes.
In that case the best course of action would be to drop these `notes` from the test.

steakhal replied:
I think there is a miscommunication.
None of the `expected-note` comments in any of your tests are effective. All of those lines are dead code.
It is because the `-analyzer-output=text` is not present in the RUN line in any of your tests.

Since the checker emits a custom note, there should be one test file testing it, or just have it enabled in all of your RUN lines in all files.

steakhal replied:
Ah I see. It may have something to do with the raw use of the addNote API, which is not usually done. We usually use BugReport visitors for putting down notes and probably their behavior is different. I'm not entirely convinced that using notes in the default analyzer output consumer is expected. Do you know any other checker that emits notes in a similar fashion - in other words, even without `-analyzer-output=`.

I tried to grep for one case when we emit notes from a static analyzer checker and could not find one.
Despite this concern, I think it's fine to have notes without enabling the text consumer. If we have complaints about this we will revisit this decision.

steakhal replied:
Makes sense then, thanks!

steakhal commented on clang/test/Analysis/store-to-immutable-basic.cpp:
```diff
@@ -0,0 +1,166 @@
+  x = 100; // No warning expected
+}
+
+// Write to non-const data through const pointer (should not warn - underlying memory is non-const)
+void test_const_ptr_to_nonconst_data() {
+  int data = 42;
+  const int *ptr = &data;
+  *(int*)ptr = 100; // No warning expected
+}
+
+// Write to non-const data through const reference (should not warn - underlying memory is non-const)
+void test_const_ref_to_nonconst_data() {
+  int data = 42;
+  const int &ref = data;
+  *(int*)&ref = 100; // No warning expected
+} 
```
Can you test the cases when `data` is actually `const`?

gamesh411 replied:
I believe that test case would be the one at the top of the file:
```
void test_write_through_const_pointer() {
  const int local_const = 10;
  int *ptr = (int*)&local_const;
  *ptr = 20; // expected-warning {{Writing to immutable memory is undefined behavior}}
}
```

steakhal replied:
It's not quite what I was looking for - my bad.
You are right, that `test_write_through_const_pointer` case already tested that case.
I was curious about the same but using references:

```c++
void test_ref_to_nonconst_data() {
  int data = 42;
  int &ref = data;
  ref = 100; // No warning expected
}

void test_ref_to_const_data() {
  const int data = 42;
  int &ref = *(int*)&data;
  ref = 100; // warn: ...
}
```

gamesh411 replied:
Nice, I have added these cases.

steakhal replied:
I think we still could not reach consensus here, and no test cases fulfill my request completely.
All writes in the tests happen through a pointer expression. Usually by taking the address of the reference then casting it to a wrong pointer type which is then dereferenced.

However, what I point to in my example is that forming a wrong reference is valid, as per C++ standard.
So this code is fine:
```
void test_ref_to_const_data() {
  const int data = 42;
  int &ref = *(int*)&data; // no UB
  // we are allowed to not use `ref`.
}
```
But then the `ref = 100` should trigger the checkBind and the checker diagnosing the issue.

steakhal replied:
Ah, I just noticed that I might have reacted too soon, as the PR was not updated, so your proposed changes were not pushed yet. nvm

steakhal commented on clang/test/Analysis/store-to-immutable-basic.cpp:
```diff
@@ -0,0 +1,166 @@
+  // expected-note@-1 {{Writing to immutable memory is undefined behavior. This memory region is marked as immutable and should not be modified}}
+}
+
+//===--- NEGATIVE TEST CASES ---===//
+// These tests should NOT trigger warnings
```
This comment acts like a spooky action at a distance.
One would only see a small part of a PR in the future modifying this file and unknowingly break this comment.
I'd rather just not partition the tests into segments, because it's really difficult to enforce.

gamesh411 replied:
I have removed this.

steakhal commented on clang/test/Analysis/store-to-immutable-basic.cpp:
```diff
@@ -0,0 +1,166 @@
+  // expected-note@-1 {{Writing to immutable memory is undefined behavior. This memory region is marked as immutable and should not be modified}}
+}
+
+// Write to const reference parameter
```
This and similar comments seem redundant to the function names attached.
Maybe we could just drop the comment?

gamesh411 replied:
Removed the redundant comments.

steakhal commented on clang/test/Analysis/store-to-immutable-basic.cpp:
```diff
@@ -0,0 +1,166 @@
+// Test basic functionality of StoreToImmutable checker
+// This tests direct writes to immutable regions without function modeling
+
+// Direct write to a const global variable
+const int global_const = 42; // expected-note {{Memory region is in immutable space}}
```
All of the global variables mentioned has an explicit initialized.
Could you add tests that only declares the global (aka. tentative definition).
```c++
const int tentative_global_const; // no explicit initializer, nor `extern` keyword.
```

gamesh411 replied:
This would not even compile in C++, so I have created 2 test files and split the test cases.
I have put those that can be expressed in C without compilation error in one test file, and put the C++ specific parts in another.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/StoreToImmutableChecker.cpp:
```diff
@@ -0,0 +1,146 @@
+      return true;
+  }
+
+  // Check if this is an ElementRegion accessing a const array
+  if (const ElementRegion *ER = dyn_cast<ElementRegion>(MR)) {
+    return isConstQualifiedType(ER->getSuperRegion(), C);
+  }
```
Should't we just go to the base region before doing anything else with `MR`?
The base region should denote the whole object, that was first seen. That's what we should really check for `constness`. But frankly, I couldn't see so far why would we allocate a const variable not in the immutable memory space.

gamesh411 replied:
I may be misunderstanding something in the memory modelling, but this part to me says if we encounter an element of an array, then we should check the array's declared type and see whether that is const-qualified.
Calling getBaseRegion on this would just get to the very base and check if that type is const. But rather, we are interested in whether any of the superregions are const along the way up the hierarchy, I presume.
Is my understanding correct in this regard, and if yes, should we account for this by actually do the walk up the memory modelling hierarchy and checking for const anywhere?

steakhal commented on clang/include/clang/StaticAnalyzer/Core/PathSensitive/MemRegion.h:
```diff
@@ -819,7 +819,7 @@ class SymbolicRegion : public SubRegion {
            s->getType()->isBlockPointerType());
     assert(isa<UnknownSpaceRegion>(sreg) || isa<HeapSpaceRegion>(sreg) ||
-           isa<GlobalSystemSpaceRegion>(sreg));
+           isa<GlobalSystemSpaceRegion>(sreg) || isa<GlobalImmutableSpaceRegion>(sreg));
```
I don't think this change is necessary.

gamesh411 replied:
Removed this.

steakhal reviewed (changes requested):
Thank you for the PR. It looks pretty solid.

gamesh411 commented on clang/lib/StaticAnalyzer/Checkers/StoreToImmutableChecker.cpp:
```diff
@@ -0,0 +1,146 @@
+
+  // Skip variable declarations and initializations - we only want to catch
+  // actual writes
+  if (isa<DeclStmt>(S) || isa<DeclRefExpr>(S))
```
Merged into one `isa<>()`.

gamesh411 committed:
Apply review suggestiongs by @steakhal

steakhal commented on clang/lib/StaticAnalyzer/Checkers/StoreToImmutableChecker.cpp:
```diff
@@ -0,0 +1,134 @@
+  if (!isConstVariable(MR, C))
+    return;
+
+  const SourceManager &SM = C.getSourceManager();
+  // Skip if this is a system macro (likely from system headers)
+  if (SM.isInSystemMacro(S->getBeginLoc()))
+    return;
```
Did you have a concrete case for justifying this branch? Could we have something in the tests for documenting it for history?

steakhal commented:
FYI please never tag anyone in commit title/message. And make sure that holds even if you push the Merge button on the GitHub UI. Otherwise the person would get pinged and spammed by emails forever.

gamesh411 committed:
[review-fix] fix test files

add fixme comments
add const reference test case
remove duplicate expected-notes and analyzer-output=text

gamesh411 committed:
[review-fix] add test case for complex memory hierarchy

gamesh411 committed:
[review-fix] remove isInSystemMacro check

gamesh411 committed:
[review-fix] add example note on string literal limitation

steakhal commented on clang/lib/StaticAnalyzer/Checkers/StoreToImmutableChecker.cpp:
```diff
@@ -0,0 +1,134 @@
+  if (isa<GlobalImmutableSpaceRegion>(MS))
+    return true;
+
+  // Check if this is a VarRegion with a const-qualified type
+  if (const VarRegion *VR = dyn_cast<VarRegion>(MR)) {
+    const VarDecl *VD = VR->getDecl();
+    if (VD && VD->getType().isConstQualified())
+      return true;
+  }
```
Thanks for naming the tests. I was more curious about the how part.

> How can a const VarRegion live outside of the immutable space?

gamesh411 committed:
[review-fix] implement hierarchical memregion handling

introduced an auxiliary function for checking non-ElementRegions
we are now checking ElementRegions by walking up the superregion-chain
a better name IMO for this function is isEffectivelyConstRegion

gamesh411 committed:
[cornercase] Lambda initialization gives a false positive in C++14 and below

these test cases assert the correct behavior, however without a fix for
detecting that we are currently initializing a lambda, these fail

steakhal reviewed (approved):
Thank you for sticking around. I think we objectively made this better.
I'm sure both of us learned something along the way.

Please merge this with the final recommendations in place. Thank you!

steakhal commented:
I realize that it only took 5 days to iron everything out. It felt longer, but in then we made it.

gamesh411 commented on clang/docs/analyzer/checkers/storetoimmutable_example.cpp:
```diff
@@ -8,6 +8,7 @@ void test_global_const() {
 }
 
 // String literal
+// NOTE: This only works in C++, not in C, as the analyzer treats string literals as non-const char arrays in C mode.
```
The const keyword was introduced in the ANSI C standard (C89/C90), so from that standard on, we should have string literals as const char arrays.
And according to the docs: the latest [``clang defaults to C17 by default``](https://clang.llvm.org/docs/UsersManual.html), so this seems to be a bug to me.

gamesh411 commented on clang/lib/StaticAnalyzer/Checkers/StoreToImmutableChecker.cpp:
```diff
@@ -29,29 +29,24 @@ class StoreToImmutableChecker : public Checker<check::Bind> {
-    const FieldDecl *FD = FR->getDecl();
-    if (FD && FD->getType().isConstQualified())
+  // Check if this is a TypedRegion with a const-qualified type
+  if (const TypedRegion *TR = dyn_cast<TypedRegion>(MR)) {
```
Fixed

gamesh411 commented on clang/lib/StaticAnalyzer/Checkers/StoreToImmutableChecker.cpp:
```diff
@@ -0,0 +1,136 @@
+  }
+
+  // Check if this is a SymbolicRegion with a const-qualified pointee type
+  if (const SymbolicRegion *SR = dyn_cast<SymbolicRegion>(MR)) {
```
Fixed

gamesh411 committed:
[format] fixed example file code formatting

gamesh411 committed:
[review-fix] don't repeat type names

gamesh411 committed:
[cornercase] fix false positive cornercase

this cornercase has a test case in a previous commit also with the tag
[cornercase]

gamesh411 commented:
@steakhal Thank you for your absolutely awesome work on this! I had a great time learning new stuff along the way!
Before I merge this, there is just a minor quirk to iron out.

I have encountered a weird corner case that leads to a false positive.
Please see the commits with tag `cornercase` for details.

The issue appears in C++14 and before (does not occur in C++17 or after) when a *const* lambda is initialised.
This case is detected in a hacky way by my last commit, but the final solution would be if the engine gave the info to the `checkBind` callback, whether this binding is an initialisation or not.
That information is available, but it is not passed forward to the callback.

gamesh411 commented on clang/lib/StaticAnalyzer/Checkers/StoreToImmutableChecker.cpp:
```diff
@@ -0,0 +1,134 @@
+    return Ty.isConstQualified();
+  }
+  return false;
+}
```
The new hierarchical handling logic supersedes this.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/StoreToImmutableChecker.cpp:
```diff
@@ -11,6 +11,7 @@
 //
 //===----------------------------------------------------------------------===//
 
+#include "clang/AST/ParentMap.h"
```
I don't think we need this include.

gamesh411 replied:
Yeah, this was part of my attempt to use the structure of the AST to detect the lambda initialisation. Removed.

steakhal commented on clang/lib/StaticAnalyzer/Checkers/StoreToImmutableChecker.cpp:
```diff
@@ -29,11 +30,59 @@ class StoreToImmutableChecker : public Checker<check::Bind> {
+  // distinguish between initialization and assignment, because this information
+  // is already available in the engine, it is just not passed to the checker
+  // API.
+  if (!isa<CXXConstructExpr>(S))
+    return false;
+
+  // We use elidable construction to detect initialization.
+  if (cast<CXXConstructExpr>(S)->isElidable())
+    return true;
+
+  return false;
```
I usually prefer this shorter syntax to express this:

```suggestion
  // We use elidable construction to detect initialization.
  const auto *ConstructExp = dyn_cast<CXXConstructExpr>(S));
  return ConstructExp && ConstructExp->isElidable();
```

NagyDonat replied:
Yep, this is a good suggestion -- I also felt that these lines could and should be simplified, but I wasn't able to find this solution immediately.

steakhal commented on clang/test/Analysis/store-to-immutable-lambda-init.cpp:
```diff
@@ -0,0 +1,10 @@
+// expected-no-diagnostics
+
+// In C++14 and before, when initializing a lambda, the statement given in the checkBind callback is not the whole DeclExpr, but the CXXConstructExpr of the lambda object.
+// FIXME: Once the API of checkBind provides more information about the statement, the checker should be simplified, and this this test case will no longer be a cornercase in the checker.
```
```suggestion
// FIXME: Once the API of checkBind provides more information about the statement, the checker should be simplified, and this test case will no longer be a cornercase in the checker.
```

gamesh411 replied:
Done

gamesh411 commented on clang/lib/StaticAnalyzer/Checkers/StoreToImmutableChecker.cpp:
```diff
@@ -0,0 +1,187 @@
+    const MemRegion *MR, CheckerContext &C) const {
+  // If the region is an ElementRegion, we need to check if any of the super
+  // regions have const-qualified type.
+  if (const auto *ER = dyn_cast<ElementRegion>(MR)) {
+    SmallVector<const MemRegion *, 8> SuperRegions;
+    const MemRegion *Current = MR;
+    const MemRegion *Base = ER->getBaseRegion();
+    while (Current != Base) {
+      SuperRegions.push_back(Current);
+      assert(isa<SubRegion>(Current));
+      Current = cast<SubRegion>(Current)->getSuperRegion();
+    }
+    SuperRegions.push_back(Base);
+    return llvm::any_of(SuperRegions, [&C](const MemRegion *MR) {
+      return isEffectivelyConstRegionAux(MR, C);
+    });
+  }
+
+  return isEffectivelyConstRegionAux(MR, C);
```
```suggestion
   while (true) {
     if (isEffectivelyConstRegionAux(MR, C))
       return true;
     if (auto *SR = dyn_cast<SubRegion>(MR))
       MR = SR->getSuperRegion();
     else
       return false;
   }
```

gamesh411 reviewed (commented):
Review suggestions resulting from consulting with @NagyDonat

NagyDonat commented on clang/docs/analyzer/checkers/storetoimmutable_example.cpp:
```diff
@@ -0,0 +1,32 @@
+// RUN: %clang_analyze_cc1 -analyzer-checker=alpha.core.StoreToImmutable %s
+
+// Global const variable
+const int global_const = 42;
```
@steakhal You are right that globals and locals are handled differently _by the checker implementation_, what we meant under "handled the same way" is the user experience. This is an example file that will be included in the documentation so we shouldn't highlight the global/local nature of variables if the user will see the same behavior (modifying `const` values or parts of them is reported) in either case.

steakhal replied:
Agreed. I misinterpreted the context.

gamesh411 replied:
I have made the simplification, and it still has an example for a global and a local variable, demonstrating that misuses of both are detected.

gamesh411 committed:
[review-fix] streamline example file

removed runline
collapsed the functions into a single one for brevity

gamesh411 committed:
[review-fix] add more C++ standard versions

gamesh411 committed:
[review-fix] streamline implementation

change warning and note message
simplify cornercase detection

gamesh411 committed:
[review-fix] support SubRegions not just ElementRegions

gamesh411 commented on clang/test/Analysis/store-to-immutable-basic.c:
```diff
@@ -0,0 +1,120 @@
+}
+
+void test_write_through_const_pointer() {
+  const int local_const = 10; // expected-note {{Memory region is in immutable space}}
```
Done

gamesh411 commented on clang/test/Analysis/store-to-immutable-lambda-init.cpp:
```diff
@@ -0,0 +1,10 @@
+// RUN: %clang_analyze_cc1 -analyzer-checker=alpha.core.StoreToImmutable -std=c++14 -verify %s
```
Added versions from C++11 - C++20

gamesh411 commented on clang/docs/analyzer/checkers/storetoimmutable_example.cpp:
```diff
@@ -0,0 +1,32 @@
+  str[0] = 'H'; // warn: Writing to immutable memory
+}
+
+// Const parameter
```
Removed

gamesh411 commented on clang/docs/analyzer/checkers/storetoimmutable_example.cpp:
```diff
@@ -0,0 +1,32 @@
+// RUN: %clang_analyze_cc1 -analyzer-checker=alpha.core.StoreToImmutable %s
```
Removed

gamesh411 commented on clang/docs/analyzer/checkers/storetoimmutable_example.cpp:
```diff
@@ -0,0 +1,32 @@
+// Global const variable
+const int global_const = 42;
+
+void test_global_const() {
```
Simplified

gamesh411 committed:
[review-fix] fix typo

gamesh411 committed:
[review-fix] delete stray whitespace

NagyDonat commented on clang/docs/analyzer/checkers/storetoimmutable_example.cpp:
```diff
@@ -0,0 +1,23 @@
+// Global const variable
```
I would prefer removing this comment and "// Const struct member" because they just state the obvious.

NagyDonat commented on clang/lib/StaticAnalyzer/Checkers/StoreToImmutableChecker.cpp:
```diff
@@ -0,0 +1,168 @@
+  // ugly hack to detect whether this construction is used to initialize a
+  // variable.
+  //
+  // FIXME: This should be eliminated once the API of checkBind would allow to
+  // distinguish between initialization and assignment, because this information
+  // is already available in the engine, it is just not passed to the checker
+  // API.
```
```suggestion
  // FIXME: This should be eliminated by improving the API of checkBind to
  // ensure that it consistently passes the `VarDecl` (instead of the
  // `CXXConstructExpr`) when the constructor call denotes the initialization
  // of a variable with a lambda.
```
As we discussed in person, the most probable solution to this corner case is slightly different from what you originally wrote here. I hope that this will become irrelevant soon (if you can restore consistency in the engine), but it is still slightly better to document this more accurately.

gamesh411 committed:
[review-fix] more elaborate notes

Now we mention in the bug message whether the target is a global
immutable memory, and thus the write is "fundamentally" bad, or it is
just touching a const qualified type (still bad, but for another
reason).

Also added notes where the MemRegion being written to does not have a
declaration. We now try to get the immediate cause of why the memory
should not be written, and we put a note there.

As a minor improvement, moved non-static member functions to free
functions wherever possible, we are not using anything from the class.

gamesh411 committed:
[review-fix] remove redundant comments from example

gamesh411 committed:
[review-fix] document our options for the fixme

gamesh411 committed:
Merge branch 'main' into store-to-immutable-checker

gamesh411 commented on clang/test/Analysis/store-to-immutable-basic.c:
```diff
@@ -0,0 +1,120 @@
+}
+
+void test_write_to_const_array() {
+  const int arr[5] = {1, 2, 3, 4, 5};
```
I have added more elaborate note implementation and tests.

NagyDonat commented on clang/docs/analyzer/checkers/storetoimmutable_example.cpp:
```diff
@@ -0,0 +1,21 @@
+  const int local_const = 42;
+  *(int *)&local_const = 43; // warn: Trying to write to immutable memory
+
+  // NOTE: The following works in C++, but not in C, as the analyzer treats
```
Perhaps replace "works" with "is reported" because "works" can be misunderstood as "is standard-compliant".

NagyDonat reviewed (approved):
LGTM, with one very minor nitpick in a comment.

gamesh411 committed:
[review-fix] clarify wording

gamesh411 committed:
fix formatting
